renku/core/management/migrations/m_0009__new_metadata_storage.py

```diff
@@ -0,0 +1,295 @@
+    if force:
+        client.remove_graph_files()
+    elif client.has_graph_files():
+        raise errors.OperationError("Graph metadata exists.")
+
+    database = Database.from_path(client.database_path)
+    initialize_database(database)
+
+    dependency_graph = DependencyGraph.from_database(database)
+    datasets_provenance = DatasetsProvenance(database)
+
+    commits = list(client.repo.iter_commits(paths=[f"{client.workflow_path}/*.yaml", ".renku/datasets/*/*.yml"]))
+    n_commits = len(commits)
+    commits = reversed(commits)
+
+    for n, commit in enumerate(commits, start=1):
+        communication.echo(f"Processing commits {n}/{n_commits} {commit.hexsha}", end="\n")
+
+        try:
+            _process_workflows(
+                client=client, commit=commit, database=database, dependency_graph=dependency_graph, remove=remove
+            )
+            _process_datasets(client=client, commit=commit, datasets_provenance=datasets_provenance)
+        except errors.MigrationError:
+            communication.echo("")
+            communication.warn(f"Cannot process commit '{commit.hexsha}' - Migration failed: {traceback.format_exc()}")
+        except Exception:
+            communication.echo("")
+            communication.warn(f"Cannot process commit '{commit.hexsha}' - Exception: {traceback.format_exc()}")
+
+        # NOTE: We need to commit in case we want to resume because we delete processed files
+        database.commit()
+
+
+def _process_workflows(
+    client: LocalClient, commit: Commit, database: Database, dependency_graph: DependencyGraph, remove: bool
+):
+    def replace_agents(activity):
+        agent = database["agents"].get(activity.association.agent.id)
+        if agent:
+            activity.association.agent = agent
+        elif activity.association.agent:
+            database["agents"].add(activity.association.agent)
+        agents = []
+        for agent in activity.agents:
+            existing_agent = database["agents"].get(agent.id)
+            if existing_agent:
+                agents.append(existing_agent)
+            elif agent:
+                agents.append(agent)
+                database["agents"].add(agent)
+
+        activity.agents = agents
+
+    for file_ in commit.diff(commit.parents or NULL_TREE, paths=f"{client.workflow_path}/*.yaml"):
+        # Ignore deleted files (they appear as ADDED in this backwards diff)
+        if file_.change_type == "A":
+            continue
+
+        path: str = git_unicode_unescape(file_.a_path)
+
+        if not path.startswith(".renku/workflow") or not path.endswith(".yaml"):
+            continue
+
+        if not (client.path / path).exists():
+            communication.warn(f"Workflow file does not exists: '{path}'")
+            continue
+
+        workflow = Activity.from_yaml(path=path, client=client)
+
+        activity_collection = ActivityCollection.from_activity(workflow, dependency_graph)
```

Comment:
Do we still need the class `ActivityCollection`? It seems that we could just replace it with a plain function (with mostly the same code as `from_activity` that returns a list.

Reply:
We still rely on it to update the new metadata when a `run/update/rerun` command is executed. I guess once we move to the new workflow metadata and use them directly for execution we can remove `ActivityCollection` and also remove it from here. We might want to keep it, however, to store `GroupedRun` executions.